File: crates/nexcore-mcp/src/params/faers.rs
//! FAERS Parameters
//! Tier: T3 (Domain-specific MCP tool parameters)
//!
//! FDA Adverse Event Reporting System (OpenFDA) parameters.

use rmcp::schemars::{self, JsonSchema};
use rmcp::serde::{Deserialize, Serialize};

/// Parameters for FAERS search
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersSearchParams {
    /// Drug name to search (generic or brand)
    #[serde(default)]
    pub drug_name: Option<String>,
    /// Adverse reaction to search (MedDRA PT)
    #[serde(default)]
    pub reaction: Option<String>,
    /// Filter to serious events only
    #[serde(default)]
    pub serious: Option<bool>,
    /// Max results (1-100)
    #[serde(default)]
    pub limit: Option<usize>,
}

/// Parameters for FAERS drug events
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersDrugEventsParams {
    /// Drug name to analyze
    pub drug_name: String,
    /// Number of top events to return (default: 20)
    #[serde(default)]
    pub top_n: Option<usize>,
    /// Minimum disproportionality threshold
    pub min_dispro: f64,
}

/// Parameters for FAERS signal check and disproportionality
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersSignalParams {
    /// Drug name
    pub drug_name: String,
    /// Target event term (MedDRA PT)
    pub event_name: String,
}

// ============================================================================
// FAERS A79 — Reporter-Weighted Disproportionality
// ============================================================================

/// Parameters for Algorithm A79: Reporter-Weighted Disproportionality.
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersReporterWeightedParams {
    /// Array of cases with reporter qualification
    pub cases: Vec<FaersReporterCase>,
    /// Minimum raw case count
    #[serde(default)]
    pub min_cases: Option<u32>,
    /// Diversity threshold
    #[serde(default)]
    pub diversity_threshold: Option<f64>,
}

/// A single case for reporter-weighted analysis.
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersReporterCase {
    /// Drug name
    pub drug: String,
    /// Event name
    pub event: String,
    /// Reporter qualification code
    pub qualification_code: String,
}

// ============================================================================
// FAERS A81 — Geographic Signal Divergence
// ============================================================================

/// Parameters for Algorithm A81: Geographic Signal Divergence.
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersGeographicDivergenceParams {
    /// Array of cases with country data
    pub cases: Vec<FaersGeographicCase>,
    /// Minimum total cases
    #[serde(default)]
    pub min_cases: Option<u32>,
    /// Minimum countries required
    #[serde(default)]
    pub min_countries: Option<usize>,
    /// Divergence ratio threshold
    #[serde(default)]
    pub divergence_threshold: Option<f64>,
    /// P-value threshold
    #[serde(default)]
    pub p_value_threshold: Option<f64>,
    /// Minimum cases per country
    #[serde(default)]
    pub min_country_cases: Option<u32>,
}

/// A single case for geographic analysis.
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersGeographicCase {
    /// Drug name
    pub drug: String,
    /// Event name
    pub event: String,
    /// Occurrence country (ISO 2-letter code)
    pub country: String,
}

// Re-export ETL and analytics types for qualified access (params::faers::FaersEtlRunParams)
pub use super::faers_analytics::{
    FaersOutcomeCase, FaersOutcomeConditionedParams, FaersPolypharmacyCase, FaersPolypharmacyDrug,
    FaersPolypharmacyParams, FaersSeriousnessCascadeParams, FaersSeriousnessCase,
    FaersSignalVelocityParams, FaersStandardPrr, FaersTemporalCase,
};
pub use super::faers_etl::*;

/// Parameters for FAERS drug comparison
#[derive(Debug, Deserialize, JsonSchema)]
#[serde(crate = "rmcp::serde")]
pub struct FaersCompareDrugsParams {
    /// First drug name
    pub drug1: String,
    /// Second drug name
    pub drug2: String,
    /// Number of events per drug (default: 15)
    #[serde(
        default,
        deserialize_with = "crate::params::serde_lenient::deserialize_option_usize_lenient"
    )]
    pub top_n: Option<usize>,
}

File: sdk/python/nexcore_sdk/faers.py
"""Typed FAERS methods for NexCoreClient.

Mixin class providing FDA Adverse Event Reporting System query methods.
Routes through /api/v1/faers/*.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .models import (
    FaersDrugEventsResult,
    FaersSearchResponse,
    FaersSignalCheckResult,
)

if TYPE_CHECKING:
    pass


class FaersMixin:
    """FAERS query and signal check methods."""

    def _get(self, path: str, params: dict[str, Any] | None = None) -> dict[str, Any]: ...
    def _post(self, path: str, json: dict[str, Any]) -> dict[str, Any]: ...

    def faers_search(
        self, query: str, *, limit: int = 25
    ) -> FaersSearchResponse:
        """Search FAERS adverse event reports.

        Args:
            query: Drug name, reaction term, or free-text query.
            limit: Max results (default 25, max 100).
        """
        data = self._get("/api/v1/faers/search", {"query": query, "limit": limit})
        return FaersSearchResponse.model_validate(data)

    def faers_drug_events(
        self, drug: str, *, limit: int = 20
    ) -> FaersDrugEventsResult:
        """Get top adverse events for a drug.

        Args:
            drug: Drug name (generic or brand).
            limit: Top N events (default 20, max 100).
        """
        data = self._get("/api/v1/faers/drug-events", {"drug": drug, "limit": limit})
        return FaersDrugEventsResult.model_validate(data)

    def faers_signal_check(
        self, drug: str, event: str
    ) -> FaersSignalCheckResult:
        """Check for disproportionality signal between a drug and event.

        Args:
            drug: Drug name.
            event: Adverse event (MedDRA preferred term).
        """
        data = self._post(
            "/api/v1/faers/signal-check", {"drug": drug, "event": event}
        )
        return FaersSignalCheckResult.model_validate(data)
